### PR TITLE
Add HarfBuzz flag to SDL2_ttf rev 703ebc7c66fd.

### LINKED ISSUE #12845: Add HarfBuzz flag to SDL2_ttf
The [current SDL_ttf port](https://github.com/emscripten-core/emscripten/pull/3728) is apparently very old (5 years?) and lacks new features.

It is currently possible to compile SDL ttf with USE_UARFBUZZ flag which would enable  more complex script support.

I'd happily work on a PR if your consider the update and perhaps point me to the current compile/port scripts, since I didn't find them anywhere. Thanks.

### PR DESCRIPTION
**Pull request by mehdisadeghi**
`TTF_USE_HARFBUZ` flag adds complex script support
to SDL_ttf, hence the update to the latest revesion.

Fixes #12845.

### CODE REVIEW

**Comment by mehdisadeghi** [comment]
I have a hard time understanding the failed tests (not the flake8!). I'd appreciate any tips here.

**Comment by kripken** [comment]
Oh, I see you've already updated the URL here to upstream directly - can ignore my comment on the PR to the ports repo then :smile:

File: tools/ports/sdl2_ttf.py [hunk 1]
```diff
@@ -5,22 +5,22 @@
...
 
 
 def get(ports, settings, shared):
-  ports.fetch_project('sdl2_ttf', 'https://github.com/emscripten-ports/SDL2_ttf/archive/' + TAG + '.zip', 'SDL2_ttf-' + TAG, sha512hash=HASH)
+  ports.fetch_project('sdl2_ttf','https://hg.libsdl.org/SDL_ttf/archive/' + TAG + '.zip', 'SDL2_ttf-' + TAG, sha512hash=HASH)
   libname = ports.get_lib_name('libSDL2_ttf')
 
   def create():
-    src_root = os.path.join(ports.get_dir(), 'sdl2_ttf', 'SDL2_ttf-' + TAG)
+    src_root = os.path.join(ports.get_dir(), 'sdl2_ttf', 'SDL_ttf-' + TAG)
```
Comment (kripken) [inline tools/ports/sdl2_ttf.py:23]: This name change may explain the error on CI. I think you need to update `embuilder.py` as well.

Comment (mehdisadeghi) [reply tools/ports/sdl2_ttf.py:23]: @kripken thanks for pointing it out. Is there anyway I can run the tests locally? Any hints would go a long way.

Comment (mehdisadeghi) [reply tools/ports/sdl2_ttf.py:23]: I installed the requirements from the requirments file and I tried `pytest tests` but I get many many errors. So I think the env is not properly set. I also checked the circle pi file but I didn't mange to replicate it locally.

Comment (kripken) [reply tools/ports/sdl2_ttf.py:23]: You should be able to run the test locally with `./tests/runner.py browser.test_sdl2_ttf`. But the part that fails is the test for the frozen cache. You can set that in your `.emscripten` file, but then you'd also need to do the other stuff the CI does.

Maybe the simplest thing is to
* Do `emcc --clear-cache --clear-ports`
* Run the embuider command that CI does.
* Run that test, and see if it mentions it builds anything - any building would be an error.

Comment (kripken) [reply tools/ports/sdl2_ttf.py:23]: (looks like the embuilder command is `./embuilder.py build ALL`)

Comment (mehdisadeghi) [reply tools/ports/sdl2_ttf.py:23]: @kripken thanks for the tips. I was able to build and run the tests as you said, however I got a strange failure in sdl2_ttf test which I can't undersand. Perhaps you have a better idea: 
```AssertionError: Unexpected difference:
--- expected
+++ actual
@@ -1 +1 @@
-/report_result?0
+/report_result?12```

Moreover, regarding the SDL2_ttf to SDL_ttf name change, I noticed the zipfile downloaded from the upstream contains a folder named `SDL_ttf` and extracts to a folder named accordingly, hence using `SDL_ttf` in the script.

**Review by kripken — APPROVED** [review]
Ok good, looks like the testing issue has been sorted out?

There was a minor merge conflict that I fixed.

Last thing, I do wonder if automatically using harfbuzz is the right thing - maybe it should be optional. But I see freetype is not optional, and I'm not sure adding an option is that important. We can always consider adding an option later. However, please add a mention in the changelog, that SDL2_ttf now uses Harfbuzz, so at least it is less surprising.

**Comment by mehdisadeghi** [comment]
@kripken thanks for the approval.

>  please add a mention in the changelog, that SDL2_ttf now uses Harfbuzz

Sure I do. Moreover, I will try to add also another test, with a new image which actually uses HarfBuzz for shaping.

**Comment by mehdisadeghi** [comment]
@kripken as you suggested I added a line to the ChangeLog and explained the change. Furthermore, I added a new test with accompanying reference image which does Arabic text shaping. A new font was necessary since the Liberation font does not contain any Arabic glyphs.

**Comment by mehdisadeghi** [comment]
> Last thing, I do wonder if automatically using harfbuzz is the right thing - maybe it should be optional.

btw, the SDL2_ttf compiled with HarfBuzz support should not affect any existing code. There are two new function calls, namely `TTF_SetDirection` and `TTF_SetScript`, which are necessary for using HarfBuzz for shaping, without calling them nothing changes (as you can see in the tests). These functions are part of the header file, however have no effect if the library is not compiled with HarfBuzz support.

File: ChangeLog.md [hunk 2]
```diff
@@ -33,6 +33,7 @@ Current Trunk
   may notice that `RelWithDebInfo` will now include debug info (it did not
   before, which appears to have been an error), and that `Release` will
   use `-O3` instead of `-O2` (which is a better choice anyhow). (#13083)
+- `SDL2_ttf` now uses upstream compiled with `TTF_USE_HARFBUZ` flag.
```
Comment (kripken) [inline ChangeLog.md:36]: This is in the wrong place - it should be on line 26, as it will be in the next release.

(I can't seem to add a suggestion on github to fix that, sorry. Maybe because the place it should go is outside of the diff region.)

File: tests/notofont/LICENSE_OFL.txt [hunk 3]
```diff
@@ -0,0 +1,92 @@
+This Font Software is licensed under the SIL Open Font License,
```
Comment (kripken) [inline tests/notofont/LICENSE_OFL.txt:1]: I think the new test should go under `tests/third_party/notofont` - is that right @sbc100 ?

File: tests/test_browser.py [hunk 4]
```diff
@@ -3129,9 +3129,16 @@ def post_build(*args):
   def test_sdl2_ttf(self):
     shutil.copy2(path_from_root('tests', 'freetype', 'LiberationSansBold.ttf'), self.get_dir())
     self.btest('sdl2_ttf.c', reference='sdl2_ttf.png',
-               args=['-O2', '-s', 'USE_SDL=2', '-s', 'USE_SDL_TTF=2', '--embed-file', 'LiberationSansBold.ttf'],
+               args=['-O2', '-s', 'USE_SDL=2', '-s', 'USE_SDL_TTF=2', '-s', 'USE_HARFBUZZ=1', '--embed-file', 'LiberationSansBold.ttf'],
```
Comment (kripken) [inline tests/test_browser.py:3132]: why is this needed - doesn't SDL2_ttf automatically add it?

File: ChangeLog.md (hunk 2 above)
Comment (mehdisadeghi) [reply ChangeLog.md]: Oh, my bad. I thought I had moved that to the right place (I had it corrected locally). Fixed.

File: tests/notofont/LICENSE_OFL.txt (hunk 3 above)
Comment (mehdisadeghi) [reply tests/notofont/LICENSE_OFL.txt]: Moved it to `third_party` folder.

File: tests/test_browser.py (hunk 4 above)
Comment (mehdisadeghi) [reply tests/test_browser.py:3132]: I'm not sure whether I undestood you correctly. Aren't these flags supposed to be passed to `emcc` when using ports? Are theses tests treated differently? (infact I removed the flag locally and the tests passed just fine)

**Comment by mehdisadeghi** [comment]
@kripken I applied the requested changes, however some (most probably) unrelated tests are failing (such as test_html5_webgl_destroy_context, test_html5_webgl_create_context_no_antialias, etc) . I tried to figure out why to no avail.

**Comment by sbc100** [comment]
Rebasing (or merging master) should fix those failures.

**Comment by kripken** [comment]
Sorry for the delay @mehdisadeghi 

Looks good aside from the test failures. As @sbc100 I think a merge will fix them. I added a merge now resolving the conflict on the changelog, so should be good now I hope.

**Comment by mehdisadeghi** [comment]
Thanks @sbc100 and @kripken, it's green now!
